docs(rfc): RFC-0004 Plan future releases (#2449)
neicnordic/sensitive-data-archive: docs/rfcs/0004-plan-releases.md

---
status: exploring
date: "2026-05-28"
authors:
   - "@jbygdell"
consulted: []
informed: []
---

# Plan future releases

## Context and Problem Statement

> Currently releases are created on every merge to the main branch without proper release notes or changelog.

Reviewer: I am probably missing some context here, but I was not sure how to read this against #2439, which merged on the morning of the 28th and moved the release workflows to `workflow_dispatch` with an explicit version input, and #2448 the following day, which made the pipeline validate the changelog rather than write it (`release_sda.yml:47-57` now bails out if it cannot parse a previous version from `sda/CHANGELOG.md`).

Was the intent that this sentence describes the situation that motivated that work, rather than where we are now? If so, would it be worth saying so explicitly, so a later reader does not take it as a description of the current pipeline?

What still seems open to me is the part you describe in the PR body, deciding in advance what actually goes into a release, which #2439 does not really address. Would it be worth reframing the context around that?

Author: This RFC where written before the changes to how the releases are done now and simply not updated.
That change in the release handling is probably in part due to the RFC.

## Decision Drivers

> * **Structured releases** - Features should be planned in advance so that the codebase doesn't contain half completed features. For example implementation of "v2" shared components should be rolled out across the board so that applicaiton config stays consistent for all apps.

Reviewer 2: ```suggestion
* **Structured releases** - Features should be planned in advance so that the codebase doesn't contain half completed features. For example implementation of "v2" shared components should be rolled out across the board so that application config stays consistent for all apps.
```

## Considered Options

> * **Make use of Epics** - Use epics to group a set if Issues into a future release. The PRs for the Issues should not be merged into main on completion but rather into a branch dedicated to the upcoming release.

Reviewer 2: ```suggestion
* **Make use of Epics** - Use epics to group a set of Issues into a future release. The PRs for the Issues should not be merged into main on completion but rather into a branch dedicated to the upcoming release.
```

> * **Make use of Epics** - Use epics to group a set of Issues into a future release. The PRs for the Issues should not be merged into main on completion but rather into a branch dedicated to the upcoming release.

Reviewer: I might be reading these wrong, but they felt to me like parts of one proposal rather than alternatives to each other, since "Make use of Epics" already prescribes the dedicated branch, and the patch/minor/major definitions below seem to hold whichever branching model we pick.

Would it be worth listing a couple of alternatives alongside them? Staying trunk based with Milestones for the scope and tags cut from `main` would be one. Something like [release-please](https://github.com/googleapis/release-please-action) could be another, although our commit subjects are only partly conventional today, so that might be a bigger change than it first looks. Even if we end up where the RFC already points, would having the rejected options written down make the eventual ADR easier to write?

> * **Use release branches** - Start using branches for each upcoming relase with the name format `Release-X.Y.Z`. Where `X.Y.Z` is the sematic version to be released.

Reviewer 2: ```suggestion
* **Use release branches** - Start using branches for each upcoming release with the name format `Release-X.Y.Z`. Where `X.Y.Z` is the semantic version to be released.
```

> * **Use release branches** - Start using branches for each upcoming release with the name format `Release-X.Y.Z`. Where `X.Y.Z` is the semantic version to be released.

Reviewer: I was not sure which version `X.Y.Z` refers to here. `release_sda.yml:152-180` seems to ship seven images under one tag driven by `sda/CHANGELOG.md`, covering sda, validator orchestrator, download, doa, postgres, rabbitmq and sftp inbox, while `sda-admin` and `charts/sda-svc` have their own workflows and changelogs.

Was the intent that a release branch covers only that train, or that the chart and the CLI would follow it too? I also noticed the branch we already have is called `release_v1` rather than a full `X.Y.Z`, so maybe the naming here is still open?

There is also `sda-cli`, which a release already makes a statement about through `app.minimalcliversion` (`sda-download/internal/config/config.go:327`). I have a vague memory of a discussion about moving it into this repo. If that happened, would it join the same train, and either way should the RFC say how the minimum client version is picked for a release?

Author: The X.Y.Z would refer to  the targeted semantic version, in most cases Z would be `0` since feature additions warrant a minor release bump.

* **Be specific on the version** - The version tag for an upcoming release should reflect what it contains.

>   * Patch - Small bug fixes, dependabot library updates and such - This is a transparent update that have no effect on configuration of a running application.

Reviewer 2: ```suggestion
  * Patch - Small bug fixes, Dependabot library updates and such - This is a transparent update that has no effect on configuration of a running application.
```

>   * Patch - Small bug fixes, Dependabot library updates and such - This is a transparent update that has no effect on configuration of a running application.

Reviewer: I could not quite tell from this whether Dependabot PRs are meant to go to the release branch, to `main`, or to both while a branch is open. Since they sit in the patch category here, I assume they should keep flowing?

I noticed we already have something close to this in `release_v1`, where `.github/dependabot.yaml` carries a second full set of package managers with `target-branch: release_v1`, from line 64 to the end of the file. If a branch is named `Release-X.Y.Z` and so changes name every release, would not that config need editing each cycle, unlike the stable `release_v1` name?

More generally, would it be worth a short pros and cons section? I was mostly looking for what a release branch costs us, since a branch held open for a while also tends to drift and integration only really gets tested at merge time.

Author: Given that we are thinking of deprecating Dependabot or at least run it very seldom it's more that kind of changes that dependabot does currently i.e. updates libraries of applications that does no see active development.

>   * Minor - Feature additions and larger bug fixes that can affect how an app is configured - Might require changes to a running applications configuration.

Reviewer 2: ```suggestion
  * Minor - Feature additions and larger bug fixes that can affect how an app is configured - Might require changes to a running application's configuration.
```

>   * Major - Breaking changes where a running application can not be rolled back to a previos version.

Reviewer 2: ```suggestion
  * Major - Breaking changes where a running application can not be rolled back to a previous version.
```

## Open Questions

> * Descide on a future release strategy

Reviewer 2: ```suggestion
* Decide on a future release strategy
```

> * Rework the RFC into an ADR

Reviewer: Since the README describes this section as the main output of an RFC, I wondered whether these two are meant as placeholders for now. "Decide on a future release strategy" reads to me as close to the title, and "Rework the RFC into an ADR" is the promotion step, so I was not sure what we would be converging on.

Would it help to put the things we do not agree on yet here instead? I was thinking of ones like: where does a hotfix branch from while a release branch is open (`release_sda.yml:22-33` already allows `hotfix/*`), how long can a release branch reasonably live, who decides the scope and who cuts the release, whether `main` stays releasable while the branch stabilises, and whether the three approvals from three partners rule applies again to the release branch merge. Or perhaps some of those are already settled and I have just missed the discussion?
